### PR TITLE
add bcrypt + simplify hash protocol

### PR DESCRIPTION
**Pull request by tannernelson**
This PR adds BCrypt hashing as well as greatly simplifies the `HashProtocol`. I believe this puts forward a much more future proof API for hashing and improves security.

### 1.0 Problems

Currently, the `HashProtocol` allows the creation of [keyed hashes](http://crypto.stackexchange.com/questions/10757/difference-between-salted-hash-and-keyed-hashing). It also supports hashers maintaining a defaultKey. This makes the logic around creating non-keyed hashes vs. keyed hashes ambiguous.
The ambiguity about what to do in these situations was made clear when attempting to implement the BCrypt hasher.

### Simplifying `HashProtocol`

By simplifying the `HashProtocol` as much as possible, we can allow the individual hashers greater flexibility in how they decide to deal with keyed hashes. I think this also signals that the `HashProtocol` will be able to adapt to new types of hashing we may want to add in the future.

### Check Function

The `HashProtocol` now includes a `check` function. This is rather trivial to implement for algorithms like SHA and MD5, but crucial for algorithms like BCrypt. Not only does this allow us to compensate for algorithms with more complex checking processes, it also provides a simpler API for the end user.

```swift
let hash = try drop.hash.make("foo")
let passes = try drop.hash.check("foo", matchesHash: hash)
```

### Using BCrypt

- Set `"bcrypt"` to key `droplet.hash` in the config files (`"crypto"` by default).
- Add a `bcrypt.json` file:

```json
{
    "workFactor": x
}
```

### Errors

Error enums have been cleaned up a bunch.

### Breaking

This PR obviously breaks API, so we will need to include it in Vapor 2.0. That said, I don't see this affecting too many users. Likely some are passing in a `key: ` value directly to the `drop.hash.make` function and will have to instead create a separate hasher for that key, like so: `CryptoHasher(..., key: key).make(...)`.

Besides that, only users using advanced features like implementing their own Hasher will need to make a lot of changes. And even then, these changes will be simplifying existing code and hopefully fairly obvious even without consulting the migration docs.

### CODE REVIEW

File: Sources/Settings/Config.swift [hunk 1]
```diff
@@ -12,6 +12,31 @@ public struct Config: NodeBacked {
     }
 }
 
+public enum ConfigError: Error {
+    case unsupported(value: String, key: [String], file: String)
+    case missing(key: [String], file: String, desiredType: Any.Type)
+    case unknown(Error)
```
Comment (LoganWright) [inline Sources/Settings/Config.swift]: let's use `open` for consistency

File: Sources/Vapor/Hash/BCryptHasher.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    /// Create a BCryptHasher using the
+    /// specified work factor.
+    ///
+    /// See BCryptHasher.workFactor
+    public init(workFactor: Int = 10) {
+        self.workFactor = workFactor
+    }
+
+    /// See HashProtocol.make
+    public func make(_ message: Bytes) throws -> Bytes {
```
Comment (LoganWright) [inline Sources/Vapor/Hash/BCryptHasher.swift]: Should this also be passing the salt back in some way to check later?

File: Sources/Vapor/Hash/BCryptHasher.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        let salt = BCryptSalt(cost: workFactor)
+        return BCrypt.hash(password: message.string, salt: salt).bytes
+    }
+
+    /// See HashProtocol.check
+    public func check(_ message: Bytes, matchesHash digest: Bytes) throws -> Bool {
```
Comment (LoganWright) [inline Sources/Vapor/Hash/BCryptHasher.swift]: Similar question here, shouldn't we need the salt passed in to verify hash

File: Sources/Vapor/Hash/CryptoHasher.swift [hunk 4]
```diff
@@ -1,113 +1,209 @@
...
+            return hash.hexString.bytes
+        case .plain:
+            return hash
+        }
     }
 
-    public enum Error: Swift.Error {
-        case hashWithoutKeyUnsupported
-        case config(String)
+    /// See HashProtocol.check
```
Comment (LoganWright) [inline Sources/Vapor/Hash/CryptoHasher.swift]: I think maybe hash protocol needs to change to account for salting stuff better, I think this forces user to use same each time, is there a way around it I'm not sure of? I haven't spent too much time w/ crypto apis.

File: Sources/Vapor/Hash/CryptoHasher.swift [hunk 5]
```diff
@@ -1,113 +1,209 @@
...
 
-        self.init(method: method, defaultKey: key)
+extension CryptoHasher.Encoding: StringInitializable {
+    public init?(from string: String) throws {
+        switch string {
+        case "hex":
+            self = .hex
+        case "base64":
+            self = .base64
+        default:
```
Comment (LoganWright) [inline Sources/Vapor/Hash/CryptoHasher.swift]: No `plain`?

File: Sources/Vapor/Hash/HashProtocol.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
+/// Creates hash digests
+public protocol HashProtocol {
+    /// Given a message, this method
+    /// returns a hashed digest of that message.
+    func make(_ message: Bytes) throws -> Bytes
```
Comment (LoganWright) [inline Sources/Vapor/Hash/HashProtocol.swift]: Here I think maybe something like `throws -> (hash: Bytes, salt: Bytes)`

File: Sources/Vapor/Hash/HashProtocol.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    /// Checks whether a given digest was created
+    /// by the supplied message.
+    ///
+    /// Returns true if the digest was created
+    /// by the supplied message, false otherwise.
+    func check(_ message: Bytes, matchesHash: Bytes) throws -> Bool
```
Comment (LoganWright) [inline Sources/Vapor/Hash/HashProtocol.swift]: Then here, something like `check(_ message: Bytes, withSalt: Bytes, matchesHash: Bytes) throws -> Bool`

File: Sources/Vapor/Hash/HashProtocol.swift [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    /// See HashProtocol.make
+    public func make(_ string: BytesConvertible) throws -> Bytes {
+        return try make(string.makeBytes())
+    }
+
+    /// See HashProtocol.check
+    public func check(_ message: BytesConvertible, matchesHash digest: BytesConvertible) throws -> Bool {
+        let message = try message.makeBytes()
+        let digest = try digest.makeBytes()
+        return try check(
```
Comment (LoganWright) [inline Sources/Vapor/Hash/HashProtocol.swift]: Is this infinite recursion?

Comment (tannernelson) [inline Sources/Vapor/Hash/HashProtocol.swift]: Nope since they are both Bytes it calls the protocol method.

Comment (LoganWright) [inline Sources/Vapor/Hash/HashProtocol.swift]: Ahhh, these are not defaults, but rather conveniences!

**Pull request merged**